winpr/libwinpr/utils/collections/BipBuffer.c

```diff
@@ -0,0 +1,435 @@
+		return -1;
+
+	block = BipBuffer_WriteTryReserve(bb, size - status, &blockSize);
+
+	if (block)
+	{
+		writeSize = size - status;
+
+		if (writeSize > blockSize)
+			writeSize = blockSize;
+
+		CopyMemory(block, &data[status], writeSize);
+		BipBuffer_WriteCommit(bb, writeSize);
+		status += writeSize;
+
+		if ((status == size) || (writeSize < blockSize))
+			return status;
+	}
+
+	block = BipBuffer_WriteTryReserve(bb, size - status, &blockSize);
+
+	if (block)
+	{
+		writeSize = size - status;
+
+		if (writeSize > blockSize)
+			writeSize = blockSize;
+
+		CopyMemory(block, &data[status], writeSize);
+		BipBuffer_WriteCommit(bb, writeSize);
+		status += writeSize;
+
+		if ((status == size) || (writeSize < blockSize))
+			return status;
+	}
+
+	return status;
+}
+
+BYTE* BipBuffer_ReadTryReserve(wBipBuffer* bb, size_t size, size_t* reserved)
+{
+	size_t reservable = 0;
+
+	if (!reserved)
+		return NULL;
+
+	if (bb->blockA.size == 0)
+	{
+		*reserved = 0;
+		return NULL;
+	}
+
+	reservable = bb->blockA.size;
+
+	if (size && (reservable > size))
+		reservable = size;
+
+	*reserved = reservable;
+	return &bb->buffer[bb->blockA.index];
+}
+
+BYTE* BipBuffer_ReadReserve(wBipBuffer* bb, size_t size)
+{
+	BYTE* block = NULL;
+	size_t reserved = 0;
+
+	if (BipBuffer_UsedSize(bb) < size)
+		return NULL;
+
+	block = BipBuffer_ReadTryReserve(bb, size, &reserved);
+
+	if (reserved == size)
+		return block;
+
+	if (!BipBuffer_Grow(bb, bb->size + 1))
+		return NULL;
+
+	block = BipBuffer_ReadTryReserve(bb, size, &reserved);
+
+	return block;
+}
+
+void BipBuffer_ReadCommit(wBipBuffer* bb, size_t size)
+{
+	if (!bb)
+		return;
+
+	if (size >= bb->blockA.size)
+	{
+		BipBlock_Copy(bb->blockA, bb->blockB);
+		BipBlock_Clear(bb->blockB);
+	}
+	else
+	{
+		bb->blockA.size -= size;
+		bb->blockA.index += size;
+	}
+}
+
+int BipBuffer_Read(wBipBuffer* bb, BYTE* data, size_t size)
+{
+	int status = 0;
+	BYTE* block = NULL;
+	size_t readSize = 0;
+	size_t blockSize = 0;
+
+	if (!bb)
+		return -1;
+
+	block = BipBuffer_ReadTryReserve(bb, 0, &blockSize);
+
+	if (block)
+	{
+		readSize = size - status;
+
+		if (readSize > blockSize)
+			readSize = blockSize;
+
+		CopyMemory(&data[status], block, readSize);
+		BipBuffer_ReadCommit(bb, readSize);
+		status += readSize;
+
+		if ((status == size) || (readSize < blockSize))
+			return status;
+	}
+
+	block = BipBuffer_ReadTryReserve(bb, 0, &blockSize);
+
+	if (block)
+	{
+		readSize = size - status;
+
+		if (readSize > blockSize)
+			readSize = blockSize;
+
+		CopyMemory(&data[status], block, readSize);
+		BipBuffer_ReadCommit(bb, readSize);
+		status += readSize;
+
+		if ((status == size) || (readSize < blockSize))
+			return status;
+	}
+
+	return status;
+}
+
+/**
+ * Construction, Destruction
+ */
+
+wBipBuffer* BipBuffer_New(size_t size)
+{
+	wBipBuffer* bb;
+
+	bb = (wBipBuffer*) calloc(1, sizeof(wBipBuffer));
+
+	if (bb)
+	{
+		SYSTEM_INFO si;
+
+		GetSystemInfo(&si);
+
+		bb->pageSize = (size_t) si.dwPageSize;
```

Comment:
The page size is not supposed to change over time, perhaps the GetSystemInfo() could be done just once and the result stored in a static variable...